### ISSUE TITLE
Passing storage options into CSV.files

### ISSUE THREAD

**Issue by dpro-shc** [intake/intake#800]
I'm getting some authentication issues using the .files method on the CSV driver (intake 0.70) when reading from Azure Blob, it seems like it's because the storage options of the catalog entry aren't getting passed into the get_fs_token_paths from fsspec's get_fs_token_paths:

`def files(self):
        if self._files is not None:
            return self._files

        urlpath = self._get_cache(self._urlpath)[0]
        if not isinstance(urlpath, list):
            urlpath = [urlpath]

        glob_in_path = any("*" in path for path in urlpath)
        if self.pattern is None and not glob_in_path:
            self._files = urlpath
        else:
            self._files = get_fs_token_paths(urlpath)[2]

        return sorted(self._files)`

Proposed change: 

`self._files = get_fs_token_paths(urlpath,storage_options=self._storage_options)[2]`

Error for reference:
`File "/usr/local/lib/python3.11/site-packages/intake/source/csv.py", line 126, in files
    self._files = get_fs_token_paths(urlpath)[2]
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/fsspec/core.py", line 646, in get_fs_token_paths
    paths = expand_paths_if_needed(paths, mode, num, fs, name_function)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/fsspec/core.py", line 576, in expand_paths_if_needed
    expanded_paths.extend(fs.glob(curr_path))
                          ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/fsspec/asyn.py", line 118, in wrapper
    return sync(self.loop, func, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/fsspec/asyn.py", line 103, in sync
    raise return_result
  File "/usr/local/lib/python3.11/site-packages/fsspec/asyn.py", line 56, in _runner
    result[0] = await coro
                ^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/fsspec/asyn.py", line 789, in _glob
    allpaths = await self._find(
               ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/adlfs/spec.py", line 928, in _find
    infos = await self._details([b async for b in blobs])
                                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/adlfs/spec.py", line 928, in <listcomp>
    infos = await self._details([b async for b in blobs])
                                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/azure/core/async_paging.py", line 142, in __anext__
    return await self.__anext__()
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/azure/core/async_paging.py", line 145, in __anext__
    self._page = await self._page_iterator.__anext__()
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/azure/core/async_paging.py", line 94, in __anext__
    self._response = await self._get_next(self.continuation_token)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/azure/storage/blob/aio/_list_blobs_helper.py", line 90, in _get_next_cb
    process_storage_error(error)
  File "/usr/local/lib/python3.11/site-packages/azure/storage/blob/_shared/response_handlers.py", line 184, in process_storage_error
    exec("raise error from None")   # pylint: disable=exec-used # nosec
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.11/site-packages/azure/storage/blob/aio/_list_blobs_helper.py", line 83, in _get_next_cb
    return await self._command(
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/azure/core/tracing/decorator_async.py", line 77, in wrapper_use_tracer
    return await func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/azure/storage/blob/_generated/aio/operations/_container_operations.py", line 1770, in list_blob_flat_segment
    map_error(status_code=response.status_code, response=response, error_map=error_map)
  File "/usr/local/lib/python3.11/site-packages/azure/core/exceptions.py", line 164, in map_error
    raise error
azure.core.exceptions.ClientAuthenticationError: Server failed to authenticate the request. Please refer to the information in the www-authenticate header.
RequestId:***
Time:2024-04-04T21:31:49.6196095Z
ErrorCode:NoAuthenticationInformation

**Comment by dpro-shc** [comment]
Update, I checked a different version of the CSV driver and found the following implementation, which works as expected. It included my suggestion:

`    def files(self):
        if self._files is not None:
            return self._files

        urlpath = self._get_cache(self._urlpath)[0]
        if not isinstance(urlpath, list):
            urlpath = [urlpath]

        glob_in_path = any("*" in path for path in urlpath)
        if self.pattern is None and not glob_in_path:
            self._files = urlpath
        else:
            protocol = split_protocol(urlpath[0])[0]
            fs, _, paths = get_fs_token_paths(urlpath, storage_options=self._storage_options)
            self._files = sorted(paths)
            if protocol:
                self._files = [fs.unstrip_protocol(fn) for fn in self._files]

        return self._files`
        
I'm not sure why the version I installed (0.7.0) has a different implementation, could it be the source of the package install?

**Comment by martindurant** [comment]
Are you willing to try Intake 2? Fixing things in the current codebase and making newer and newer releases is easier than fixing past releases.

**Comment by dpro-shc** [comment]
Yeah, I am. But not until I get comfortable with it enough to migrate the drivers we I depend on to 2.0. I've been checking every couple of weeks for changes, but I still haven't reached the same level of comfort as I did from <2.0 honestly.

That being said, I think there should be a legacy repo for the newest pre-2.0 release because not only would it be useful for reference purposes, but also to track these kinds of changes on the <2.0 versions so while 2.0 matures, users could still get these fixes.

Is the most recent release 0.7? I think I had seen a .9 somewhere... In the version I'm looking at, there's a simple fix for a bug in include_path_column but I can't tell whether such issues were addressed or not and having a reference would help me identify that.

**Comment by martindurant** [comment]
0.7.0; with most functionality for a builtin driver like CSV, you should be able to use it with v2 also. Your existing extra drivers may well work too, but let me know if not. Some very specific functionality was totally removed, like intake caching and filepatterns.

**Comment by dpro-shc** [comment]
Noted.

Unfortunately I use a custom Excel driver and the Salesforce one, neither are builtin. I never used caching, but was definitely using some of the file pattern functionality, see my comments on for my specific usage [issue 776](https://github.com/intake/intake/issues/776). I'll have to test them out sometime when I reach the same level of comfort with 2.0

I still stand by the point that there should be a release between .7 and 2.0 for reference and for containing these kinds of simple fixes like the aforementioned one and this one in the ._read_pandas:
```

if drop_path_column:
            df_part = df_part.drop([path_column], axis=1)
else:
            self._csv_kwargs['include_path_column'] = True

```
where the change is the addition of the else block

**Comment by dpro-shc** [comment]
I see there's a PandasExcel Reader, that makes me a whole lot less concerned. But I don't see what the V2 analog of catalog_entry.files() is, what happened there?

**Comment by martindurant** [comment]
The readers are greatly simplified. So you if you want details about the files that would be used given a glob pattern, you would need a different reader, for example
```
class OpenFilesreader(FileReader):
    url_arg = "urlpath"
    implements = {datatypes.FileData}
    func = "fsspec:open_files"
    output_instance = "fsspec.core:OpenFiles"
```
runs
```
>>> path =  "./**/*.py"
>>> data = intake.datatypes.FileData(path)
>>> data.to_reader(reader="Open").read()
<List of 49 OpenFile instances>
```

**Comment by martindurant** [comment]
(I'll add this to the codebase - maybe as FSSpecReader?)

**Comment by dpro-shc** [comment]
Ah I see. Aren't most Readers compatible with FSSpec? How would the V1 functionality work for say a CSV case where we have a folder that we want to concat and read? Would it be a pipeline including the FSSpecReader and PandasCSVReader?

**Comment by dpro-shc** [comment]
Ah I think I get it now. A feature I wanted in V1 was the ability to change the "default engine" for both read and pipeline from dask to say pandas, duckdb, etc. With V2 it's possible to do so? I see that duckdb also supports globbing so I could do DuckCSV -> DuckToPandas as well to get the V1 functionality?

**Comment by martindurant** [comment]
Yes, you could equally use duck to read the data and convert to pandas (if you want). You could also have pipelines data->duck->pandas, data->dask->pandas side by side in a catalog. If you only had one of these (the default) but at runtime decided you wanted to do something different with the data description, that's OK too, perhaps with [auto_pipeline](https://intake.readthedocs.io/en/latest/api2.html#intake.readers.convert.auto_pipeline) if you don't know what intermediate hop you need.
